internal/controller/clusterorder_controller.go

```diff
@@ -20,46 +20,247 @@ package controller
+	return []component{
+		{"Namespace", r.newNamespace},
+		{"ServiceAccount", r.newServiceAccount},
+		{"RoleBinding", r.newAdminRoleBinding},
+	}
+}
+
 // ClusterOrderReconciler reconciles a ClusterOrder object
 type ClusterOrderReconciler struct {
 	client.Client
 	Scheme *runtime.Scheme
 }
 
+func newClusterReference() *cloudkitv1alpha1.ClusterOrderClusterReferenceType {
+	return &cloudkitv1alpha1.ClusterOrderClusterReferenceType{}
+}
+
 // +kubebuilder:rbac:groups=cloudkit.openshift.io,resources=clusterorders,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=cloudkit.openshift.io,resources=clusterorders/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups=cloudkit.openshift.io,resources=clusterorders/finalizers,verbs=update
+// +kubebuilder:rbac:groups="",resources=namespaces;serviceaccounts,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=rolebindings,verbs=get;list;watch;create;update;patch;delete
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the ClusterOrder object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.19.0/pkg/reconcile
-//
-//nolint:revive
 func (r *ClusterOrderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	_ = ctrllog.FromContext(ctx)
 
-	// TODO(user): your logic here
+	instance := &cloudkitv1alpha1.ClusterOrder{}
+	err := r.Client.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
 
-	return ctrl.Result{}, nil
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.handleUpdate(ctx, req, instance)
+	} else {
+		return r.handleDelete(ctx, req, instance)
+	}
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *ClusterOrderReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	labelSelector := cloudkitClusterOrderNameLabel
+
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&cloudkitv1alpha1.ClusterOrder{}).
+		Watches(
+			&corev1.Namespace{},
+			handler.EnqueueRequestsFromMapFunc(r.mapObjectToCluster),
+			builder.WithPredicates(labelPredicate(labelSelector)),
+		).
+		Watches(
+			&corev1.ServiceAccount{},
+			handler.EnqueueRequestsFromMapFunc(r.mapObjectToCluster),
+			builder.WithPredicates(labelPredicate(labelSelector)),
+		).
+		Watches(
+			&rbacv1.RoleBinding{},
+			handler.EnqueueRequestsFromMapFunc(r.mapObjectToCluster),
+			builder.WithPredicates(labelPredicate(labelSelector)),
+		).
 		Complete(r)
 }
+
+func labelPredicate(label string) predicate.Predicate {
```

Comment:
You could use this instead: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/predicate#LabelSelectorPredicate